### PR TITLE
Bootstrap refactor to C

### PR DESCRIPTION
**Pull request by bauerm97**
Fixes #

This PR is a merge of the entire bootstrap-v2 process into C code. Bootstrap-v1 will still run the same way (for the most part), since it is deprecated and no longer receiving updates (we could perhaps add a parser that can read v1 formatted bootstrap files, but probably not worth it). This currently does not remove any of the old code, but rather implements the new code and bypasses the old code. After thorough testing we will be able to eliminate the old code that is not being used.

After this code is merged, we will be able to look forward to the next set of features that we can now add that were otherwise difficult to implement. Features discussed with @vsoch and @gmkurtzer include layering of images, ability to `singularity import` or `singularity export` without root, better support for docker layer caching, etc... All of these features and more will now be much easier to implement with bootstrapping occurring in C.

This should not be merged right away, but rather this needs thorough testing to ensure that all functionality is preserved and that there are no discrepancies in the new bootstrapping code. More unit tests are needed for the final release version using this code, as the current testing only does two bootstrap commands. Currently all tests pass and I haven't found any problematic cases, but I'm only one person.

Changes proposed in this pull request

 - Changes `libexec/cli/bootstrap.exec` to call `simage bootstrap`
 - Changes `src/simage.c` to recognize bootstrap option
 - Creates new `src/lib/image/bootstrap/` directory, housing bootstrap definition parsing logic, the main bootstrap logic, and the individual modules (docker, debootstrap, etc...)
 - Adds `src/lib/image/bootstrap/bootdef_parser.c`: logic for parsing a definition file
 - Adds `src/lib/image/bootstrap/bootstrap.c`: main bootstrapping logic and useful functions
 - Users can now specify %environment in definition file to create a /environment file in container
 - Adds `src/lib/image/bootstrap/*/*.c`: individual module functions, currently call on related module (python for docker, shell module located at `libexec/bootstrap/modules-v2/build-*.sh` for rest)
 - Updates `src/lib/file/file.c` to handle creation of necessary files at bootstrap time
 - Adds `src/lib/file/entrypoint/entrypoint.c` to add /.run, /.exec, and /.shell into container root filesystem
 -  Adds more unit testing for bootstrapping to `test.sh.in`; **this needs more work I think**

@singularityware-admin

### CODE REVIEW

**Comment by bauerm97** [comment]
@gmkurtzer 

Questions for you: 

- `ldconfig -r "$SINGULARITY_ROOTFS" >/dev/null 2>&1` is this line necessary in postbootstrap.sh? I just noticed this and I never implemented it in the bootstrapping code
- Currently postbootstrap.sh passes `-e -x` to /bin/sh when running the % scripts from the bootdef file, is this necessary? It seems to work fine without it
- I've omitted the `if [ "$UID" == "0" ]; then` checks from around the scripts, I assumed that since it's not an SUID binary we don't need to check. If the user makes it that far as non-root, they still won't have any privileges. Thoughts?

To do list:

- [x] Copy `%test` script into container rootfs (Oops, I glanced over this)
- [ ] Implement more robust unit tests
- [ ] Implement the `ldconfig` line (?)
- [x] Change entrypoint.c to always use `#!/bin/sh`
- [ ] Merge all bootstrap modules (except docker) into one general case
- [ ] Clean up superfluous include lines
- [x] Move `etc/defaults` -> `libexec/defaults`

**Comment by gmkurtzer** [comment]
I'm not sure about the necessisity of ldconfig running outside the rootfs anymore. But I'll review it a bit more.

The -e option tells the shell to bomb out if any errors are encountered with the scriptlets and the -x just runs the scriptlets verbosely. I like these options. 

I agree that omitting the UID check is reasonable. We can always add it back if necessary. 

Hopefully later today I will be going through this PR and will report more later. 

Thanks!

File: src/lib/file/entrypoint/entrypoint.c [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    char *rootfs_path = singularity_rootfs_dir();
+    char *dest_path = joinpath(rootfs_path, strjoin("/.", entrypoint_name));
+    char *entrypoint = filecat(strjoin(LIBEXECDIR "/singularity/defaults/", entrypoint_name));
+
+    if ( is_file(joinpath(rootfs_path, "/bin/bash")) == 0 ) {
+        helper_shell = strdup("#!/bin/bash");
+    } else {
+        helper_shell = strdup("#!/bin/sh");
+    }
+
```
Comment (gmkurtzer) [inline src/lib/file/entrypoint/entrypoint.c]: looking for `/bin/bash` isn't a good idea as these scripts will be run in the containers not the host. We should just always assume and use the POSIX required `/bin/sh`.

File: src/lib/image/bootstrap/arch/arch.c [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+#include <fcntl.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/mount.h>
+#include <sys/wait.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <sched.h>
```
Comment (gmkurtzer) [inline src/lib/image/bootstrap/arch/arch.c:32]: There is a lot of unnecessary includes, but I too am bad at cleaning them all up... But I'm good at pointing when others do the same! lol

Comment (bauerm97) [reply src/lib/image/bootstrap/arch/arch.c:32]: Sorry, I literally just copy and paste the includes. Maybe the static analysis tool Brian was using can help us clean these up? Sonar I think it is called

File: src/lib/file/entrypoint/entrypoint.c (hunk 1 above)
Comment (bauerm97) [reply src/lib/file/entrypoint/entrypoint.c]: OK. We do this currently in the bootstrapping code (postbootstrap.sh) so I wasn't sure about it.

File: src/lib/image/bootstrap/busybox/busybox.c [hunk 3]
```diff
@@ -0,0 +1,45 @@
+/*
```
Comment (gmkurtzer) [inline src/lib/image/bootstrap/busybox/busybox.c:1]: We should be able to consolidate these bootstrap modules into a single module that will look for a predefine path pattern (e.g. /usr/libexec/singularity/bootstrap/modules/%s.sh) so we don't need to create a new C module for every shell based counterpart.

Comment (bauerm97) [reply src/lib/image/bootstrap/busybox/busybox.c:1]: Is there any reason why we would want to port the individual modules into C, or are they fine being left in shell? I was thinking that we might want to port some code into C and therefore need the different module.c files. Your call, I wasn't sure

File: src/util/util.c [hunk 4]
```diff
@@ -166,7 +186,7 @@ void chomp(char *str) {
         for ( i = 1; i < len; i++ ) {
 	    str[i-1] = str[i];
 	}
-	str[len] = '\0';
+	str[len-1] = '\0';
```
Comment (gmkurtzer) [inline src/util/util.c]: Good catch!

File: test.sh.in [hunk 5]
```diff
@@ -178,15 +181,15 @@ stest 0 sudo singularity exec -w "$CONTAINER" touch /writetest.pass
...
-stest 0 sudo singularity bootstrap "$CONTAINER"
-stest 0 singularity exec "$CONTAINER" test -f /environment
-stest 0 sudo singularity exec -w "$CONTAINER" rm /environment
-stest 1 singularity exec "$CONTAINER" test -f /environment
-stest 0 sudo singularity bootstrap "$CONTAINER"
-stest 0 singularity exec "$CONTAINER" test -f /environment
-stest 0 singularity exec "$CONTAINER" test -f /.shell
-stest 0 singularity exec "$CONTAINER" test -f /.exec
-stest 0 singularity exec "$CONTAINER" test -f /.run
+stest 1 sudo singularity bootstrap "$CONTAINER"
```
Comment (gmkurtzer) [inline test.sh.in:190]: Did you mean to comment these out?

File: src/lib/file/entrypoint/entrypoint.c (hunk 1 above)
Comment (gmkurtzer) [reply src/lib/file/entrypoint/entrypoint.c]: Yep, postbootstrap runs on the host. These will run in the containers.

File: src/lib/image/bootstrap/busybox/busybox.c (hunk 3 above)
Comment (gmkurtzer) [reply src/lib/image/bootstrap/busybox/busybox.c:1]: If we port them to C, they would almost entirely be made up of shell commands, so I would argue there isn't much of a need to. Do you concur?

File: src/lib/image/bootstrap/arch/arch.c (hunk 2 above)
Comment (gmkurtzer) [reply src/lib/image/bootstrap/arch/arch.c:32]: No worries. As I said, I am guilty of that one too! lol

File: src/lib/image/bootstrap/busybox/busybox.c (hunk 3 above)
Comment (bauerm97) [reply src/lib/image/bootstrap/busybox/busybox.c:1]: Yep this was my thinking as well, but I had already created the folders before I decided to look into those files and see what was going on ;) I'll fix it up

File: src/util/util.c (hunk 4 above)
Comment (bauerm97) [reply src/util/util.c]: This should probably be merged separately and ASAP, I think it's possible to attempt to write that null byte out of allocated memory.

Comment (gmkurtzer) [reply src/util/util.c]: Done: [master 3994040]

**Comment by gmkurtzer** [comment]
note: I've fixed the conflicting files issue which resulted in cherry picking the util.c updates in my local working/testing branch.

File: src/lib/image/bootstrap/busybox/busybox.c (hunk 3 above)
Comment (gmkurtzer) [reply src/lib/image/bootstrap/busybox/busybox.c:1]: I think I'm still waiting on this one (unless I missed it)... Should I do it, or do you want to?

Comment (bauerm97) [reply src/lib/image/bootstrap/busybox/busybox.c:1]: Your call, I didn't have time to get to it in the office today, and I won't have more time until 3 or 4 hours from now. If you want to I won't say no

Comment (gmkurtzer) [reply src/lib/image/bootstrap/busybox/busybox.c:1]: ok, I'll take care of it. I also may reorganize some of the bootstrap code (I don't like it under src/lib/image as it breaks the nomenclature <-> directory standard I've been using). For example: singularity_bootstrap_yum() should be in lib/bootstrap/yum. I will take care of that too but if you find yourself with extra time, and bored, and this hasn't been committed yet, ping me in Slack. It is likely I'll get interrupted. lol

**Comment by vsoch** [comment]
@bauerm97 is this ready for merging into my branch to work on https://github.com/singularityware/singularity/pull/376 or should I wait for the checks to pass / you to tell me it's ready to go?

**Comment by bauerm97** [comment]
I'm not sure why the tests aren't passing, but I think Greg's recent commit to master broke the build. Everything here works except the Python commit Greg pushed. Sorry

**Comment by vsoch** [comment]
Oh yes, I think I just fixed the python ones! Is [this one](https://travis-ci.org/singularityware/singularity/builds/181475264#L1809) python related?

**Comment by bauerm97** [comment]
@vsoch @gmkurtzer I rebased again on top of master, and it's passing locally. This should be good to go for you

**Comment by bauerm97** [comment]
OK, I have no freaking idea why this isn't working properly. I *just* tested this locally and it was working fine

**Comment by vsoch** [comment]
cool! I have a meeting now but can work on this later in the evening - @bauerm97 quick question - I see that we are checking for the environment file, and then adding a default if necessary. What was the final decision we came on for https://github.com/singularityware/singularity/pull/376 - I thought we were going to create a rootfs/.env folder for these files to be ordered, and then I could write a docker one to it. I did a quick search on the changes in the code and I didn't see anything other than `environment` - could you point me to where I can look to see how the gears are working?

Back in a bit!

**Comment by vsoch** [comment]
Hey don't worry we will figure this out! Sometimes when I can't see a bug it helps to go away for a while and do something else or relax.  Then come back with fresh eyes and boum there he is!

**Comment by vsoch** [comment]
@gmkurtzer note that the not passing test is related to the fact that the "tag" is gone from the custom (gcr.io) image manifest obtained with the docker code (eg look at  https://gcr.io/v2/tensorflow/tensorflow/manifests/latest) there is no tag anymore. I think we should perhaps merge the PR that has updates to obtaining the manifest https://github.com/singularityware/singularity/pull/386 if the bug endures, I'll fix it, and if not, we should be good to go. Going back to sleep for now - I have a meeting at 10am Pacific and will be back to work on this around noonish! :zzz:

**Comment by vsoch** [comment]
Let's rebase and give this another go!

**Comment by bauerm97** [comment]
@vsoch Rebased and fingers crossed. This should work, as it works in my local branch and I fixed why it was breaking earlier (due to my code)

**Comment by gmkurtzer** [comment]
I have it checked out locally. Initially I've had to fix the singularity.spec.in to include the libexec/defaults into the RPM. So far things are working, I will take a closer look at teh code tomorrow.

Great job everyone!

**Comment by vsoch** [comment]
Awesome! Thanks @gmkurtzer !

**Comment by gmkurtzer** [comment]
This has been merged, not sure why GitHub is still showing this as open. I will give it a bit and then possibly have to close by hand.

**Comment by vsoch** [comment]
Did you fork and merge the fork?

**Comment by vsoch** [comment]
Yes, it's merged. I'll go ahead and close.

**Comment by gmkurtzer** [comment]
It is odd, previously GitHub would catch those and automatically mark merged... Hrmmmm, thanks for closing!

**Comment by vsoch** [comment]
I think it might be the checkbox functionality they added? Maybe things don't close if boxes are unchecked!

**Pull request closed without merging**